File: 1_ML/ml_utils.py
import math
import os


"""
K-Nearest Neighbors (KNN) algorithm.

Individual distances:
 - Manhattan
 - Euclidea

Once the algorithm finalize, an interface shows the assignment of:
    Initial population (at the left) 
    Final population (at the right)
"""

"""
Calculates de Euclidean distance of 
    two points in a 'd' dimensional space.

Args:
    a (float[]) : Point in a d-dimensional space.
    b (float[]) : Point in a d-dimensional space.
"""
def euclidean_distance(a, b):
    ret=0
    d=len(a)
    for i in range(d):
        ret+=(a[i]-b[i])**2    

    return math.sqrt(ret)

"""
Calculates de Manhattan distance of 
    two points in a 'd' dimensional space.

Args:
    a (float[]) : Point in a d-dimensional space.
    b (float[]) : Point in a d-dimensional space.
"""
def manhattan_distance(a, b):
    ret=0
    d=len(a)
    for i in range(d):
        ret+=abs(a[i]-b[i])

    return math.sqrt(ret)


"""
Read a population of individuals with 'd' variables.

The format of the file has to be as follows:
[_,..,_], [_,..,_], ... ,[_,..,_]
(All in 1 line.)

Args:
    file_name (string) : Name of the file that is going to be readed.

Return:
    ret (float[][]) : Individuals of the population.
"""
def read_population(file_name, algorithm, d):    
    ret   =[]            # return list
    point =[]            # individual    
    
    # get the root directory of the proyect
    dir=os.getcwd()       
    while(dir[-9:]!='ReCode-AI'):
        dir=os.path.dirname(dir)

    # name of the file
    if file_name==None: file_name=input('Introduce a file name: ')    
    path=os.path.join(dir,'.Others','files',algorithm, file_name+'.txt')

    # read the line    
    try:
        with open(path, 'r') as file:
            data=file.read()
    except FileNotFoundError:
        print("The file '{}' doesnt exits.".format(file_name+".txt"))
    

    # removes '[', ']'. And divides the file by ','
    data=data.replace('[', '').replace(']', '').split(', ') 
    
    # store the individuals in the return list     
    for i in range(0, len(data), d):
        point=[]
        # read and store all the variables of the individual
        for j in range(d):
            point.append(float(data[i+j]))
        
        ret.append(point)
              
    
    return ret


"""
Read an assignment of individuals.

The format of the file has to be as follows:
_ _ ... _
(All in 1 line.)

Args:
    file_name (string) : Name of the file that is going to be readed.

Return:
    ret (int[]) : Categories of each individual.
"""
def read_assignment(file_name):
    ret=[] # return list

    # get the root directory of the proyect
    dir=os.getcwd()       
    while(dir[-9:]!='ReCode-AI'):
        dir=os.path.dirname(dir)

    # name of the file
    if file_name==None: file_name=input('Introduce a file name: ')    
    path=os.path.join(dir,'.Others','files','clusters','assignment', file_name+'.txt')
    
    # read the line
    try:
        with open(path, 'r') as file:
            data=file.read()
    except FileNotFoundError:
        print("The file '{}' doesnt exits.".format(file_name+".txt"))

    data=data.split(' ')                               
    for num in data:
        ret.append(int(num))
    
    
    
    return ret

File: 1_ML/ml_utils.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Globalization;


namespace ML_Utilities
{
    public class Functions
    {
        /*
        Read a population of individuals with 'd' variables.

        The format of the file has to be as follows:
        [_,..,_], [_,..,_], ... ,[_,..,_]
        (All in 1 line.)

        Args:
            fileName (string)  : Name of the file that is going to be readed.
            algorithm (string) : Type of directory.
            d (int)            : Number of variables per individual.

        Return:
            ret (List<double[]>) : Individuals of the population.
        */
        public static List<double[]> ReadPopulation(string fileName, string algorithm,  int d) 
        {
            var ret=new List<double[]>();             // return list    

            try 
            {
                // get the root directory of the proyect
                string dir=Directory.GetCurrentDirectory();
                while (!dir.EndsWith("ReCode-AI")) 
                {
                    dir=Directory.GetParent(dir).FullName;
                }
                
                // name of the file
                if (fileName==null) fileName=Console.ReadLine();
                string path=Path.Combine(dir, ".Others", "files", 
                                        algorithm, fileName+".txt");
                
                // read the line 
                string data;
                data=File.ReadAllText(path); 

                
                // Remove brackets and split by "], ["
                string[] sets = data.Trim('[', ']').Split(new string[] { "], [" }, StringSplitOptions.RemoveEmptyEntries);

                // Parse each set of numbers
                foreach (string set in sets)
                {
                    // split by commas to get individual string values
                    string[] values=set.Split(',');

                    // Convert string values to double and store them in an array
                    double[] numbers = new double[values.Length];

                    for (int i = 0; i < values.Length; i++)
                    {
                        numbers[i] = double.Parse(values[i], CultureInfo.InvariantCulture);
                    }

                    // Add the array to the list
                    ret.Add(numbers);
                }

            }
            catch (Exception ex)
            {
                Console.WriteLine("An error occurred: " + ex.Message);
            }
            
            

            

            return ret;
        }
    }
}